wagtail/models/sites.py

```diff
@@ -209,42 +212,58 @@ def get_site_root_paths():
         - `root_url` - The scheme/domain name of the site (for example 'https://www.example.com/')
         - `language_code` - The language code of the site (for example 'en')
         """
-        result = cache.get("wagtail_site_root_paths")
-
-        # Wagtail 2.11 changed the way site root paths were stored. This can cause an upgraded 2.11
-        # site to break when loading cached site root paths that were cached with 2.10.2 or older
-        # versions of Wagtail. The line below checks if the any of the cached site urls is consistent
-        # with an older version of Wagtail and invalidates the cache.
-        if result is None or any(len(site_record) == 3 for site_record in result):
-            result = []
-
-            for site in Site.objects.select_related(
-                "root_page", "root_page__locale"
-            ).order_by("-root_page__url_path", "-is_default_site", "hostname"):
-                if getattr(settings, "WAGTAIL_I18N_ENABLED", False):
-                    result.extend(
-                        [
-                            SiteRootPath(
-                                site.id,
-                                root_page.url_path,
-                                site.root_url,
-                                root_page.locale.language_code,
-                            )
-                            for root_page in site.root_page.get_translations(
-                                inclusive=True
-                            ).select_related("locale")
-                        ]
-                    )
-                else:
-                    result.append(
+
+        # Reuse previously generated value if the method was already called
+        # for the current thread
+        result = getattr(_root_paths_cache, "value", None)
+        if result is not None:
+            # Return a copy to prevent accidental pollution of cached values
+            return deepcopy(result)
```

Comment:
I guess another key benefit of the deepcopy here is that we prevent references to objects in the cache, 
which is important for enabling garbage collection / preventing memory leakage.